artemis-server/src/main/java/org/apache/activemq/artemis/core/replication/ReplicationManager.java

```diff
@@ -551,49 +580,52 @@ private void sendLargeFile(AbstractJournalStorageManager.JournalContent content,
-            // through ActiveMQBuffer class leaving this buffer free to be reused on the next copy
-            while (true) {
-               final ByteBuf buffer = PooledByteBufAllocator.DEFAULT.directBuffer(size, size);
-               buffer.clear();
-               ByteBuffer byteBuffer = buffer.writerIndex(size).readerIndex(0).nioBuffer();
-               final int bytesRead = channel.read(byteBuffer);
-               int toSend = bytesRead;
-               if (bytesRead > 0) {
-                  if (bytesRead >= maxBytesToSend) {
-                     toSend = (int) maxBytesToSend;
-                     maxBytesToSend = 0;
-                  } else {
-                     maxBytesToSend = maxBytesToSend - bytesRead;
-                  }
-               }
-               logger.debug("sending " + buffer.writerIndex() + " bytes on file " + file.getFileName());
-               // sending -1 or 0 bytes will close the file at the backup
-               // We cannot simply send everything of a file through the executor,
-               // otherwise we would run out of memory.
-               // so we don't use the executor here
-               sendReplicatePacket(new ReplicationSyncFileMessage(content, pageStore, id, toSend, buffer), true);
-               packetsSent++;
-
-               if (packetsSent % flowControlSize == 0) {
-                  flushReplicationStream(action);
+         raf = new RandomAccessFile(file.getJavaFile(), "r");
+         fileChannel = raf.getChannel();
+         while (true) {
+            long chunkSize = Math.min(size, fileSize - offset);
+            int toSend = (int) chunkSize;
+            if (chunkSize > 0) {
+               if (chunkSize >= maxBytesToSend) {
+                  toSend = (int) maxBytesToSend;
+                  maxBytesToSend = 0;
+               } else {
+                  maxBytesToSend = maxBytesToSend - chunkSize;
                }
-               if (bytesRead == -1 || bytesRead == 0 || maxBytesToSend == 0)
-                  break;
             }
+            logger.debug("sending " + toSend + " bytes on file " + file.getFileName());
+            // sending -1 or 0 bytes will close the file at the backup
+            // We cannot simply send everything of a file through the executor,
+            // otherwise we would run out of memory.
+            // so we don't use the executor here
+            sendSyncFileMessage(new ReplicationSyncFileMessage(content, pageStore, id, raf, fileChannel, offset, toSend), offset + toSend == fileSize);
+            packetsSent++;
+            offset += toSend;
+
+            if (packetsSent % flowControlSize == 0) {
+               flushReplicationStream(action);
+            }
+            if (toSend == 0 || maxBytesToSend == 0)
+               break;
          }
          flushReplicationStream(action);
+
+      } catch (Exception e) {
+         if (fileChannel != null)
+            fileChannel.close();
```

Comment:
And raf? 
Maybe would be better to mark the expected last packet and make netty able to register a listener and let it close it on deallocate (after a successful/failed write)?

Reply:
You mean on caller side we add a callback where is called in netty channelFutureListener after write?

Reply:
Yes, but just the supposed last packet need that listener, while the other writes can just use a void promise instead